syntax = "proto3";
option go_package = "github.com/IRATI/ipcm_api";

package ipcm_api;

// The greeting service definition.
service IPCM {
    rpc ListIpcps (ListIpcpsReq) returns (ListIpcpsRes);
    rpc GetIfaceForDif (GetIfaceForDifReq) returns (GetIfaceForDifRes);
    rpc QueryRib (QueryRibReq) returns (QueryRibRes);
    rpc CreateIpcp (CreateIpcpReq) returns (CreateIpcpRes);
    rpc DestroyIpcp (DestroyIpcpReq) returns (DestroyIpcpRes);
    rpc EnrollToDif (EnrollToDifReq) returns (EnrollToDifRes);
    rpc AssignToDif (AssignToDifReq) returns (AssignToDifRes);
}

// IPCP info for list-ipcps
message IPCP {
    uint64 id = 1;
    string name = 2;
    string type = 3;

    // DIF name, if state = 4
    string difname = 4;

    // State number
    uint32 state = 5;

    // Interpreted state value
    string state_string = 6;

    repeated string apps = 7;
    repeated int32 flows = 8;
}

// Query RIB object
message RIBObject {
    string name = 1;
    string class = 2;
    int64 instance = 3;
    string displayable_value = 4;
}

// list-ipcps

message ListIpcpsReq {}

message ListIpcpsRes {
    repeated IPCP ipcp = 1;
}

// get-iface-for-dif

message GetIfaceForDifReq {
    string dif = 1;
}

message GetIfaceForDifRes {
    string iface = 1;
}

// query-rib

message QueryRibReq {
    uint64 ipcp_id = 1;
    string object_class = 2;
    string object_name = 3;
}

message QueryRibRes {
    repeated RIBObject rib_object = 1;
}

// create-ipcp

message CreateIpcpReq {
    string ipcp_type = 1;
    string dif_name = 2;
}

message CreateIpcpRes {
    uint64 ipcp_id = 1;
    string name = 2;
}

// destroy-ipcp

message DestroyIpcpReq {
    int32 ipcp_id = 1;
}

message DestroyIpcpRes {}

// enroll-to-dif

message EnrollToDifReq {
    uint64 ipcp_id = 1;
    string dif_name = 2;
    string supporting_dif_name = 3;
    string neighbor_process_name = 4;
    string neighbor_process_inst = 5;
}

message EnrollToDifRes {
    int32 ms_delay = 1;
}

// assign-to-dif

message AssignToDifReq {
    uint64 ipcp_id = 1;
    string dif_name = 2;
    string dif_template = 3;
}

message AssignToDifRes {}

